Add coverage checker for I18n locales

benthorner opened the pull request:
https://trello.com/c/s122Fg91/959-we-dont-monitor-unused-and-untested-i18n-translations

This adds a GOVUKI18nCov class* that generates a coverage report of the
I18n keys used after a script (e.g. rspec) has finished running. I've
removed a couple of unused locales that the report has identified as
being unused - the remainder are used but not covered by the tests.

I'm not really sure where this code should live, but it does seem to be 
beneficial to monitor coverage of our locales :-).

*there are other tools out there, such as i18n-tasks, but I couldn't find one 
that produces a good quality results for our use case - we have a lot of 
dynamic uses of I18n in our code, as well as simple uses in the views.

kevindew commented on spec/support/govuk_i18n_cov.rb:
```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+
+class GOVUKI18nCov
```
It'd be more conventional for this to be capitalised as Govuk. Though I wonder if it needs a Govuk prefix as it's not specific to us.

benthorner replied:
Fair enough - I was being cautious :-).

kevindew commented on spec/support/govuk_i18n_cov.rb:
```diff
@@ -0,0 +1,81 @@
+
+      @started = true
+      @used_keys = []
+      ::I18n::Backend::Simple.send :include, I18n::Registry
```
Looks like include is a public method: https://github.com/ruby-i18n/i18n/blob/0c5dab494d9b043e00662d8e789229c33045c024/lib/i18n/backend/simple.rb#L24

kevindew commented on spec/support/govuk_i18n_cov.rb:
```diff
@@ -0,0 +1,81 @@
+    end
+  end
+
+  class << self
```
Note, operating on the eigenclass was discouraged in the GOV.UK style guide before it was replaced. Hence it's seen very little in GOV.UK code. I don't mind it myself, though I wonder if using a Singleton pattern might make it a bit easier to follow as a class.

benthorner replied:
Fair enough - SimpleCov seems to be doing this, but I like it better after moving to the Singleton pattern.

kevindew commented on spec/support/govuk_i18n_cov.rb:
```diff
@@ -0,0 +1,81 @@
+  private
+
+    def get_all_i18n_keys
+      ymls = Dir.glob("config/locales/**/*.yml")
```
Should it use Rails to access these? We configure this here: https://github.com/alphagov/content-publisher/blob/57e6867e9033cef6c2c24f88e1ded8d7569de61e/config/application.rb#L32

Though after looking this up I see that it will include all the gems too so this might be better.

benthorner replied:
You got it in one! I've hopefully made this a bit clearer, plus making it somewhat configurable to support testing.

kevindew commented on spec/support/govuk_i18n_cov.rb:
```diff
@@ -0,0 +1,81 @@
+    def write_report(unused)
+      FileUtils.mkdir_p("coverage")
+
+      File.open("coverage/i18n", "w") do |f|
```
I have a recollection that you said this file was in the same format as simplecov, hopefully I'm not making that up. I notice this is a list but simplecov drops out a HTML file.

kevindew commented on spec/support/govuk_i18n_cov.rb:
```diff
@@ -0,0 +1,81 @@
+
+class GOVUKI18nCov
+  module I18n
+    module Registry
```
Any reason this should be two modules deep?

benthorner replied:
Nope - you find the flaws in my pasted code ;-). I've revised it now, which I think has made it a bit clearer too!

kevindew commented on spec/support/govuk_i18n_cov.rb:
```diff
@@ -0,0 +1,81 @@
+    end
+
+    def normalise_keys(keys)
+      keys.uniq.map(&:to_s).map { |s| s.sub(/^\./, "") }
```
Just out of curiosity, what happens that would mean a key needs normalising?

benthorner replied:
It's possible to use I18n in a number of ways, which include using scoped and symbolic arguments, as well as specifying the optional root '.' element. That said, it looks like that's not a problem for our in-app keys, so I've removed it for now :-).

kevindew commented on spec/support/i18n_cov.rb:
```diff
@@ -0,0 +1,80 @@
+  end
+
+  def write_report(unused)
+    FileUtils.mkdir_p("coverage")
```
Did you mean for this to be relative to Rails.root too?

kevindew commented on spec/support/i18n_cov.rb:
```diff
@@ -0,0 +1,80 @@
+
+  def get_all_i18n_keys
+    locales.flat_map { |locale, ts| key_chains_for(locale, ts) }
+      .map { |key_chain| key_chain.drop(1).join(".") }
```
This might need some documentation - refactor, comment, or commit message - to explain the drop of the locale part

benthorner commented:
@kevindew thanks for taking another look. I've done a bit more refactoring: added some constants with Rails.join, renamed some of the methods, made the scoping of which used keys we care about more prominent, and made the drop(1) a concern of the key_chains_for method, which hopefully makes it clearer, given the updated commit message :-).

benthorner commented:
@kevindew ah, I hadn't thought about 'default' like that - it was meant to refer to it's usage in generating the default locales they aren't specified in '.start'. I've changed it to be 'in_app', which reflects the method it's used in :-).
